### DISCUSSION TITLE
Hoping for some guidance on SSL

### DISCUSSION THREAD

**Discussion by strunker** [miguelgrinberg/microdot#123] [Q&A]
So microdot is great, but it is super large for my needs. I was hoping to write my own socket code and wrap it in SSL socket. MIcrodots ssl implementation works perfectly fine, I am just trying to emulate it in my own code.

Both of the below bodies of code suffer from the same issue. I tried to use the ssl context class to wrap an ssl socket. Then on my laptop im trying to hit the esp webserver. Anytime I try I get the below error.  Hoping that maybe @miguelgrinberg or someone else has run into this and can perhaps offer some guidance on how I can resolve this. Thanks!

> requests.exceptions.SSLError: HTTPSConnectionPool(host='10.1.1.206', port=8443): Max retries exceeded with url: / (Caused by SSLError(SSLError(1, '[SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:2607)')))

```
import machine
from microdot_ssl import create_ssl_context
import ssl
import socket
try:
    addr = socket.getaddrinfo('0.0.0.0', 8443)[0][-1]
    s = socket.socket()
    s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    s.bind(addr)
    s.listen(5)
except:
    pass

print('listening on', addr)

cert="cert.der"
key="key.der"
with open(cert, 'rb') as f:
    cert = f.read()
with open(key, 'rb') as f:
    key = f.read()

while True:
    cl, addr = s.accept()
    print('client connected from', addr)
    print("1")
    sslctx = create_ssl_context(cert, key)
    sslctx.wrap_socket(cl, server_side=True)
    print("2")
    cl.send('HTTP/1.1 200 OK\n')
    cl.send('Content-Type: text/html\n')
    cl.send('Connection: close\n\n')
    cl.close()

```
```
import machine
import ssl
import socket
try:
    addr = socket.getaddrinfo('0.0.0.0', 8443)[0][-1]
    s = socket.socket()
    s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    s.bind(addr)
    s.listen(5)
except:
    pass

print('listening on', addr)

cert="cert.der"
key="key.der"
with open(cert, 'rb') as f:
    cert = f.read()
with open(key, 'rb') as f:
    key = f.read()

while True:
    cl, addr = s.accept()
    print('client connected from', addr)
    print("1")
    ssl.wrap_socket(cl, server_side=True,cert=cert,key=key)
    print("2")
    cl.send('HTTP/1.1 200 OK\n')
    cl.send('Content-Type: text/html\n')
    cl.send('Connection: close\n\n')
    cl.close()

```

**Comment by miguelgrinberg** [comment]
I can't really provide advice, but the SSL implementation in MicroPython is tricky. I suggest you look at how I'm wrapping the socket and do the same. The `SSLContext` class that I wrote is extremely simplistic and is only designed to cover my own needs, it is not designed to for general use. If you want to use it, you have to make sure you use it exactly as I do, since other usages are likely not going to work.

**Comment by strunker** [comment]
I noticed that ha. The syntax is a bit rigid, so are the cert requirements, etc. I did ultimately end up sorting it out with the help of Damien George. Very thankful for the guidance. I missed a small piece of syntax and my socket was actually not properly being wrapped in ssl, which led to the bizarre errors I was experiencing. More info below if you are at all curious. In any event, thank you for the reply here, and great work on microdot. If I needed something full featured I would for sure use it. My needs for this particular project are much lighter weight. Thanks!

https://github.com/micropython/micropython/issues/11208#event-8941793257
